supersonic/cursor/infrastructure/row_hash_set.cc

```diff
@@ -455,6 +459,78 @@ void RowHashSetImpl::FindInternal(
   }
 }
 
+template <bool HASH_COMP_ONLY>
+void RowHashSetImpl::InsertUniqueFromIterator(ViewRowIterator& iterator,
+   const bool_const_ptr selection_vector, FindResult* result) {
+
+
+    while (iterator.next()) {
+      const int64 query_row_id = iterator.current_row_index();
+
+      rowid_t* const result_row_id = (result) ? result->mutable_row_ids() + query_row_id : NULL;
+
+      if (selection_vector != NULL && !selection_vector[query_row_id]) {
+        if (result_row_id) {
+          *result_row_id = kInvalidRowId;
+        }
+      } else {
+        // TODO(user): The following 12 lines are identical to the piece
+        // of code in FindInternal. Needs refactoring (but be careful about a
+        // performance regression).
+        int hash_index = (hash_mask_ & query_hash_[query_row_id]);
+        int index_row_id = last_row_id_[hash_index];
+
+        int *prev = &last_row_id_[hash_index];
+
+        do {
+          if (index_row_id == -1) {
```

Comment:
Despite all the other work (CityHash), LRU buckets, moving the HASH_COMP_ONLY branch out (template arg). The end of the linked list check is the largest time sink. Not sure what we can do about this branch.